configtx/application.go

```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
+			return nil
+		}
+	}
+
+	msp.IntermediateCerts = append(msp.IntermediateCerts, cert)
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// RemoveIntermediateCert removes an trusted intermediate certificate trusted by the organization MSP.
+func (m *ApplicationOrgMSP) RemoveIntermediateCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	certs := msp.IntermediateCerts[:]
+	for i, c := range msp.IntermediateCerts {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
+
+	msp.IntermediateCerts = certs
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddOUIdentifier adds a fabric organizational unit for the organization MSP.
+func (m *ApplicationOrgMSP) AddOUIdentifier(ou membership.OUIdentifier) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	for _, o := range msp.OrganizationalUnitIdentifiers {
+		if reflect.DeepEqual(o, ou) {
+			return nil
+		}
+	}
+
+	msp.OrganizationalUnitIdentifiers = append(msp.OrganizationalUnitIdentifiers, ou)
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// RemoveOUIdentifier removes a fabric organizational unit for the organization MSP.
+func (m *ApplicationOrgMSP) RemoveOUIdentifier(ou membership.OUIdentifier) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	ous := msp.OrganizationalUnitIdentifiers[:]
+	for i, o := range msp.OrganizationalUnitIdentifiers {
+		if reflect.DeepEqual(o, ou) {
+			ous = append(ous[:i], ous[i+1:]...)
+			break
+		}
+	}
+
+	msp.OrganizationalUnitIdentifiers = ous
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// SetCryptoConfig sets the configuration for the cryptographic algorithms for the organization MSP.
+func (m *ApplicationOrgMSP) SetCryptoConfig(cryptoConfig membership.CryptoConfig) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	msp.CryptoConfig = cryptoConfig
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddTLSRootCert adds an TLS root certificate trusted by the organization MSP.
+func (m *ApplicationOrgMSP) AddTLSRootCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	for _, c := range msp.TLSRootCerts {
+		if c.Equal(cert) {
+			return nil
+		}
+	}
+
+	msp.TLSRootCerts = append(msp.TLSRootCerts, cert)
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// RemoveTLSRootCert removes an trusted TLS root certificate from the organization MSP.
+func (m *ApplicationOrgMSP) RemoveTLSRootCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	certs := msp.TLSRootCerts[:]
+	for i, c := range msp.TLSRootCerts {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
```

Comment:
I checked the `msp.validateCACerts` function and it doesn't look like we validate TLS intermediate certs against the TLS root certs. @wlahti was this an oversight when we originally added the check for regular `RootCerts`/`IntermediateCerts`? https://github.com/hyperledger/fabric-config/blob/master/configtx/msp.go#L481-L517

Reply:
We may want to still call `validateCACerts` here like my comment above for `RemoveRootCert` and in a followup story fix the above if `TLSIntermediateCerts` should also be validated against the `TLSRootCerts`

Reply:
added `validateCACerts()`

Reply:
> I checked the `msp.validateCACerts` function and it doesn't look like we validate TLS intermediate certs against the TLS root certs. @wlahti was this an oversight when we originally added the check for regular `RootCerts`/`IntermediateCerts`? https://github.com/hyperledger/fabric-config/blob/master/configtx/msp.go#L481-L517

Yes, it looks like this was an oversight, likely due to me not being very specific in the JIRA task and specifying we should validate the root/intermediate CA certs for both regular and TLS CA certificates. Good catch. :)